openedx/core/lib/api/parsers.py

```diff
@@ -1,7 +1,90 @@
-from rest_framework import parsers
+"""
+Custom DRF request parsers.  These can be used by views to handle different
+content types, as specified by `<Parser>.media_type`.
 
+To use these in an APIView, set `<View>.parser_classes` to a list including the
+desired parsers.  See http://www.django-rest-framework.org/api-guide/parsers/
+for details.
+"""
 
-class MergePatchParser(parsers.JSONParser):
+from django.utils.translation import ugettext as _
+
+from rest_framework.exceptions import ParseError, UnsupportedMediaType
+from rest_framework.parsers import FileUploadParser, JSONParser
+
+
+class TypedFileUploadParser(FileUploadParser):
+    """
+    Handles upload of files, ensuring that the media type is supported, and
+    that the uploaded filename matches the Content-type.
+
+    Requirements:
+        * The view must have an `upload_media_types` attribute which is a
+          set (or other container) enumerating the mimetypes of the supported
+          media formats
+
+          Example:
+
+              View.upload_media_types = {'audio/mp3', 'audio/ogg', 'audio/wav'}
+
+        * Content-type must be set to a supported type (as
+          defined in View.upload_media_types above).
+
+          Example:
+
+              Content-type: audio/ogg
+
+        * Content-disposition must include a filename with a valid extension
+          for the specified Content-type.
+
+          Example:
+
+              Content-disposition: attachment; filename="lecture-1.ogg"
+    """
+
+    media_type = '*/*'
+
+    # Add more entries to this as needed.  All extensions should be lowercase.
+    file_extensions = {
+        'image/gif': {'.gif'},
+        'image/jpeg': {'.jpeg', '.jpg'},
+        'image/pjpeg': {'.jpeg', '.jpg'},
+        'image/png': {'.png'},
+        'image/svg': {'.svg'},
+    }
+
+    def parse(self, stream, media_type=None, parser_context=None):
+        """
+        Parse the request, returning a DataAndFiles object with the data dict
+        left empty, and the body of the request placed in files['file'].
+        """
+
+        upload_media_types = getattr(parser_context['view'], 'upload_media_types', None)
+        if upload_media_types is not None and media_type not in upload_media_types:
+            raise UnsupportedMediaType(media_type)
+
+        filename = self.get_filename(stream, media_type, parser_context)
+        if media_type in self.file_extensions:
```

Comment:
So if `media_type` is not in `self.file_extensions`, we simply upload it as a `FileUploadParser` would do? Since this parser explicitly validates the mimetypes and file extensions of its uploads, I'd expect it to throw an error in that case.